src/accumulator/proof.rs

```diff
@@ -0,0 +1,310 @@
+    ///   
+    ///   let mut hashes = Vec::new();
+    ///   for i in test_values {
+    ///     let mut engine = bitcoin_hashes::sha256::Hash::engine();
+    ///      engine.input(&[i]);
+    ///      let hash = sha256::Hash::from_engine(engine);
+    ///     hashes.push(hash);
+    ///   }
+    ///
+    ///   s.modify(&hashes, &vec![]);
+    ///   let p = Proof::new(targets, proof_hashes);
+    ///   assert!(p.verify(&vec![hashes[0]] , &s).expect("No error should happens"));
+    ///```
+    pub fn verify(&self, del_hashes: &Vec<sha256::Hash>, stump: &Stump) -> Result<bool, String> {
+        let calculated_roots = self.create_root_candidates(del_hashes, stump)?;
+        let stump_roots = util::num_roots(stump.leafs);
+        let mut number_matched_roots = 0;
+
+        for i in 0..stump_roots {
+            if calculated_roots[number_matched_roots] == stump.roots[stump.roots.len() - (i + 1)] {
+                number_matched_roots += 1;
+                if number_matched_roots == calculated_roots.len() {
+                    break;
+                }
+            }
+        }
+
+        if calculated_roots.len() != number_matched_roots && calculated_roots.len() != 0 {
+            return Ok(false);
+        }
+        Ok(true)
+    }
+
+    pub fn create_root_candidates(
+        &self,
+        del_hashes: &Vec<sha256::Hash>,
+        stump: &Stump,
+    ) -> Result<Vec<sha256::Hash>, String> {
+        // Where all the root hashes that we've calculated will go to.
+        let total_rows = util::tree_rows(stump.leafs);
+        let mut targets = self.targets.clone();
+
+        // Where all the parent hashes we've calculated in a given row will go to.
+        let mut calculated_root_hashes =
+            Vec::<sha256::Hash>::with_capacity(util::num_roots(stump.leafs) as usize);
+
+        // As we calculate nodes upwards, it accumulates here
+        let mut nodes: Vec<_> = self
+            .targets
+            .to_owned()
+            .into_iter()
+            .zip(del_hashes.to_owned())
+            .collect();
+
+        let mut row = 0;
+        let mut proof_hashes_index = 0 as usize;
+
+        while row <= total_rows {
+            let res = util::detect_row_hashes(&targets, row, total_rows);
+
+            if let Ok((start, end)) = res {
+                let mut i = start as usize;
+
+                // Iterates over a row
+                while i <= end as usize {
+                    if nodes.len() == 0 {
+                        return Err(String::from("Malformed tree"));
+                    }
+
+                    // If the current position is a root, we add that to our result and don't go any further
+                    if util::is_root_position(nodes[i].0, stump.leafs, total_rows) {
+                        calculated_root_hashes.push(nodes[i].1);
+                    } else if nodes.len() > i + 1 && (nodes[i].0 | 1 == nodes[i + 1].0) {
```

Comment:
`nodes[i].0 | 1` is confusing for someone who's not familiar with the code. Could we make a `is_right_niece()` in utils and call that function?